reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2728 @@
+        }
+        else
+        {
+            ASSERT(SGList->SgElementCount == 0);
+
+            TD->HwTD.gTD.CurrentBuffer = 0;
+            TD->HwTD.gTD.BufferEnd = 0;
+
+            TD->TransferLen = 0;
+        }
+
+        PrevTD = TD;
+
+        TD = OHCI_AllocateTD(OhciExtension, OhciEndpoint);
+        OhciTransfer->PendingTDs++;
+
+        PrevTD->HwTD.gTD.NextTD = TD->PhysicalAddress;
+        PrevTD->NextHcdTD = (ULONG)TD;
+    }
+    while (TransferedLen < TransferParameters->TransferBufferLength);
+
+    if (TransferParameters->TransferFlags & USBD_SHORT_TRANSFER_OK)
+    {
+        PrevTD->HwTD.gTD.Control.BufferRounding = TRUE;
+        OhciTransfer->Flags |= OHCI_TRANSFER_FLAGS_SHORT_TRANSFER_OK;
+    }
+
+    PrevTD->HwTD.gTD.Control.DelayInterrupt = OHCI_TD_INTERRUPT_IMMEDIATE;
+    PrevTD->HwTD.gTD.NextTD = TD->PhysicalAddress;
+    PrevTD->NextHcdTD = (ULONG)TD;
+
+    TD->HwTD.gTD.NextTD = 0;
+    TD->NextHcdTD = 0;
+
+    OhciTransfer->NextTD = TD;
+    OhciEndpoint->HcdTailP = TD;
+
+    OhciEndpoint->HcdED->HwED.TailPointer = TD->PhysicalAddress;
+
+    OHCI_EnableList(OhciExtension, OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+OHCI_SubmitTransfer(IN PVOID ohciExtension,
+                    IN PVOID ohciEndpoint,
+                    IN PVOID transferParameters,
+                    IN PVOID ohciTransfer,
+                    IN PVOID sgList)
+{
+    POHCI_EXTENSION OhciExtension = ohciExtension;
+    POHCI_ENDPOINT OhciEndpoint = ohciEndpoint;
+    PUSBPORT_TRANSFER_PARAMETERS TransferParameters = transferParameters;
+    POHCI_TRANSFER OhciTransfer = ohciTransfer;
+    PUSBPORT_SCATTER_GATHER_LIST SGList = sgList;
+    ULONG TransferType;
+
+    DPRINT_OHCI("OHCI_SubmitTransfer: ... \n");
+
+    RtlZeroMemory(OhciTransfer, sizeof(OHCI_TRANSFER));
+
+    OhciTransfer->TransferParameters = TransferParameters;
+    OhciTransfer->OhciEndpoint = OhciEndpoint;
+
+    TransferType = OhciEndpoint->EndpointProperties.TransferType;
+
+    if (TransferType == USBPORT_TRANSFER_TYPE_CONTROL)
+    {
+        return OHCI_ControlTransfer(OhciExtension,
+                                    OhciEndpoint,
+                                    TransferParameters,
+                                    OhciTransfer,
+                                    SGList);
+    }
+
+    if (TransferType == USBPORT_TRANSFER_TYPE_BULK ||
+        TransferType == USBPORT_TRANSFER_TYPE_INTERRUPT)
+    {
+        return OHCI_BulkOrInterruptTransfer(OhciExtension,
+                                            OhciEndpoint,
+                                            TransferParameters,
+                                            OhciTransfer,
+                                            SGList);
+    }
+
+    return MP_STATUS_FAILURE;
+}
+
+MPSTATUS
+NTAPI
+OHCI_SubmitIsoTransfer(IN PVOID ohciExtension,
+                       IN PVOID ohciEndpoint,
+                       IN PVOID transferParameters,
```

Comment:
... and the same here for SubmitIsoTransfer?

Reply:
Changed.